lib/ProductOpener/Store.pm

```diff
@@ -268,24 +268,39 @@ sub retrieve ($file) {
 # Serializes an object in our preferred object store, removing it from legacy storage if it is present
 sub store_object ($path, $ref, $delete_old = 1) {
 	my $new_path = $path . '.json';
-	if (open(my $OUT, ">", $new_path)) {
-		# Get an exclusive lock on the file
-		# We could also lock the STO file here but it adds a small overhead and we don't
-		# tend to run old and new versions in parallel so shouldn't be an issue
-		flock($OUT, LOCK_EX);
-		my $json = $json_for_objects->encode($ref);
-		# Strip out any nul characters as many parsers can't cope with these
-		# This doesn't seem to add too much overhead
-		$json =~ s/\000//g;
-		print $OUT $json;
-		close($OUT);
+	# If the file already exists then we need to first open it non-destructively so that
+	# other code doesn't read an empty file before we have written the data
+	my $READ_LOCK;
```

Comment:
Interesting: is this not required for STO files because an empty file in that format would be considered invalid (and cause an error) in the read process? (or some other reason?)

Reply:
For sto files we use the lock_store function which handles all of the locking for us, we need to do explicit locking for JSON as there isn't an equivalent function in the JSON module. The reason it is a bit convoluted to do it in Perl code is because flock only works once you have opened a file, but opening a file for writing will empty that file so there is a small window between us opening a the file (and hence clearing it out) and applying the lock where another process might try to read the file.

Reply:
Could/should we deploy the locking change around the existing `.sto` files, independently of the JSON migration, to confirm that it is reliable?

Reply:
It doesn't work the same. `lock_store` does all of the file management / locking internally.

Reply:
Understood - my line of thinking was that we could deploy the new locking mechanism independently and before the JSON migration; it would involve removing the use of `lock_store`.

And to explain that: I am anxious about deploying two changes (1. file locking change, 2. data format change) that I think are complicated and risky within a single release, because I think it could make it more difficult to identify problems and/or rollback.

Reply:
I don't believe we are changing the locking behaviour. lock_store will basically lock the file, write to it and then release the lock. The only reason that we have to do it in a slightly more convoluted way for JSON is that we have to open the file for reading before we can put a lock on it, we can then re-open for writing to ensure nothing accesses the file part way through the write process.

Reply:
I've changed the locking to open in append mode, lock and then truncate, the same as lock_store. Also reviewed the error handling to be consistent, this being that store_object will die on error (as I think that's what lock_store would do) but retrieve_object will just carp as the existing retrieve wrapper in Store.pm catches and carps lock_retrieve errors.

Reply:
Although I'm not confident that the logging semantics are the same between this and the Perl Storable implementation -- the updated series of file-related syscalls (open, flock) does look good to me, thank you :+1: 

Are you planning to re-run the benchmarking process?  I am wary that these and perhaps other recent changes could have invalidated the previous results.